### PR TITLE
Pull request list

### LINKED ISSUE #2643: Pull request CI status
Show PRs CI status in the pull request list.

Designs TBD. Probably dependent on https://github.com/desktop/desktop/issues/2642.

### LINKED ISSUE #2780: Create a PR List Component
We should be able to see all of our open PRs in a list like structure.

See issue #2642 for more context

### PR DESCRIPTION
**Pull request by joshaber**
screen shot 2017-09-22 at 2 36 58 pm
screen shot 2017-09-22 at 3 38 20 pm
screen shot 2017-09-22 at 3 38 35 pm

Fixes #2780
Fixes #2643

Stubbing this out by doing the API requests inline. Once #2779 is done we can switch over to pulling from the proper store.

- [x] List PRs
- [x] Show CI status
- [x] Show count on the tab
- [x] Checkout the PR branch
- [x] Loading state
- [x] No PRs state
  - [x] On the default branch
  - [x] On a feature branch
- [x] Ensure it's preview-flagged thoroughly.

### CODE REVIEW

File: app/src/ui/branches/pull-request-list-item.tsx [hunk 1]
```diff
@@ -27,8 +27,8 @@ export class PullRequestListItem extends React.Component<
   public render() {
     const timeAgo = moment(this.props.created).fromNow()
     const title = this.props.title
-    const subtitle = `#${this.props.number} opened ${timeAgo} by ${this.props
-      .author}`
+    const author = this.props.author
```
Comment (j-f1) [inline app/src/ui/branches/pull-request-list-item.tsx]: `const { author, number } = this.props`?

Comment (joshaber) [reply app/src/ui/branches/pull-request-list-item.tsx]: Good point, I always forget that's a thing!

File: app/src/ui/branches/index.tsx [hunk 2]
```diff
@@ -89,7 +95,7 @@ export class Branches extends React.Component<IBranchesProps, IBranchesState> {
         <span className="pull-request-tab">
           {__DARWIN__ ? 'Pull Requests' : 'Pull requests'}
 
-          <span className="count">{count}</span>
+          {countElement}
```
Comment (j-f1) [inline app/src/ui/branches/index.tsx]: `{this.props.pullRequests && <span className="count">{this.props.pullRequests.length}</span>}`? `false` is not rendered in React.

Comment (joshaber) [reply app/src/ui/branches/index.tsx]: It's probably just a taste thing, but that's always seemed a bridge too far for me.

File: app/src/ui/branches/pull-requests-loading.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        renderItem={this.renderItem}
+        invalidationProps={groups}
+        filterDisabled={true}
+      />
+    )
+  }
+
+  private renderItem = (item: IFilterListItem) => {
+    return (
+      <div className="pull-request-loading-item">
```
Comment (j-f1) [inline app/src/ui/branches/pull-requests-loading.tsx]: How about a `style={{ opacity: (FacadeCount - item.id)/FacadeCount }}` to make them fade out towards the bottom?

**Comment by joshaber** [comment]
@niik could you hook me up with that sweet illustration you're using in the blank slate views?

File: app/src/ui/branches/pull-requests-loading.tsx (hunk 3 above)
Comment (joshaber) [reply app/src/ui/branches/pull-requests-loading.tsx]: Perhaps! @niik already has in mind a wave/pulse animation for these.

**Comment by joshaber** [comment]
I'm gonna open some issues for some polish items to be done separately.

This is ready for review. 🙏

**Comment by iAmWillShepherd** [comment]
The store and DB are just about ready, need to add a test or two. Do you wanna merge that into this branch or do you prefer me to merge it in mine?

**Comment by joshaber** [comment]
I think we can land them both and then do the refactor to use the store separately.

File: app/src/ui/toolbar/branch-dropdown.tsx [hunk 4]
```diff
@@ -48,10 +77,99 @@ export class BranchDropdown extends React.Component<IBranchDropdownProps, {}> {
...
+            repository.name,
+            pr.head.sha
+          )
+
+          pullRequestsWithStatus.push({
+            ...pr,
+            state,
+            created: new Date(pr.created_at),
+          })
+        } catch (e) {
```
Comment (iAmWillShepherd) [inline app/src/ui/toolbar/branch-dropdown.tsx]: Do we care about logging whatever this error is?

File: app/src/ui/toolbar/branch-dropdown.tsx [hunk 5]
```diff
@@ -48,10 +77,99 @@ export class BranchDropdown extends React.Component<IBranchDropdownProps, {}> {
...
+          pullRequestsWithStatus.push({
+            ...pr,
+            state: 'pending',
+            created: new Date(pr.created_at),
+          })
+        }
+      }
+
+      return pullRequestsWithStatus
+    } catch (e) {
```
Comment (iAmWillShepherd) [inline app/src/ui/toolbar/branch-dropdown.tsx]: Same with previous comment.

File: app/src/ui/branches/no-pull-requests.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        {this.renderCallToAction()}
+      </div>
+    )
+  }
+
+  private renderCallToAction() {
+    if (this.props.isOnDefaultBranch) {
+      return (
+        <div className="call-to-action">
+          Would you like to{' '}
```
Comment (iAmWillShepherd) [inline app/src/ui/branches/no-pull-requests.tsx]: I think you can use a `span` or `p` here so we won't have free text inside a div...

File: app/src/ui/branches/no-pull-requests.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+/** The placeholder for when there are no open pull requests. */
+export class NoPullRequests extends React.Component<INoPullRequestsProps, {}> {
+  public render() {
+    return (
+      <div className="no-pull-requests">
+        <img src={BlankSlateImage} className="blankslate-image" />
+
+        <div className="title">You're all set!</div>
```
Comment (iAmWillShepherd) [inline app/src/ui/branches/no-pull-requests.tsx]: Personal pref, but I think text should be wrapped in a `p`

File: app/src/ui/branches/index.tsx [hunk 8]
```diff
@@ -57,22 +73,138 @@ export class Branches extends React.Component<IBranchesProps, IBranchesState> {
...
+              ? repo.gitHubRepository.fullName
+              : repo.name
+            const isOnDefaultBranch =
+              this.props.defaultBranch &&
+              this.props.currentBranch &&
+              this.props.defaultBranch.name === this.props.currentBranch.name
+            return (
+              <NoPullRequests
+                repositoryName={name}
+                isOnDefaultBranch={!!isOnDefaultBranch}
```
Comment (iAmWillShepherd) [inline app/src/ui/branches/index.tsx]: Why the double `!!` here?

File: app/src/ui/toolbar/branch-dropdown.tsx (hunk 4 above)
Comment (joshaber) [reply app/src/ui/toolbar/branch-dropdown.tsx]: Nah, this is just filler code until we use the store.

File: app/src/ui/branches/index.tsx (hunk 8 above)
Comment (joshaber) [reply app/src/ui/branches/index.tsx]: `isOnDefaultBranch` is `boolean | null`, but the `isOnDefaultBranch` prop is `boolean`. So the double negation gets rid of the `null`.

File: app/src/ui/branches/no-pull-requests.tsx (hunk 7 above)
Comment (joshaber) [reply app/src/ui/branches/no-pull-requests.tsx]: I don't think it matters. We use `div` in lots of places.

File: app/src/ui/branches/no-pull-requests.tsx (hunk 6 above)
Comment (joshaber) [reply app/src/ui/branches/no-pull-requests.tsx]: Same as above, I don't think it matters?

**Pull request merged**